controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts

```diff
@@ -178,215 +127,392 @@ export class MetricsDashboardRepository {
-    };
+    } as any;
   }
 
   /**
    * Get latency metrics
    */
   public async getLatencyMetrics({ range = 24, endDate = getEndDate(), params }: GetMetricsProps) {
-    const prange = getPRange(range);
-
-    // get p95 of requests in last 5min
-    const p95 = this.client.query({
-      query: `histogram_quantile(0.95, sum(rate(cosmo_router_http_request_duration_milliseconds_bucket{${this.getQueryLabels(
-        params,
-      )}}[${range}h])) by (le))`,
-    });
+    const granule = getGranularity(range);
+
+    const [start, end] = getDateRange(endDate, range);
+    const [prevStart, prevEnd] = getDateRange(endDate, range, range);
+
+    const queryLatency = (quantile: string, start: number, end: number) => {
+      return this.chClient.queryPromise<{ value: number }>(`
+        WITH
+          toDateTime('${start}') AS startDate,
+          toDateTime('${end}') AS endDate
+        SELECT quantile(0.5)(value) as value FROM (
+          SELECT
+            func_rank(${quantile}, BucketCounts) as rank,
+            func_rank_bucket_lower_index(rank, BucketCounts) as bucketLowerIndex,
+            func_rank_bound(rank, bucketLowerIndex, BucketCounts, ExplicitBounds) as bound,
+            func_histogram_v2(
+                    rank,
+                    bucketLowerIndex,
+                    bound,
+                    BucketCounts,
+                    ExplicitBounds
+                ) as value,
+
+            -- Histogram aggregations
+            sumForEach(BucketCounts) as BucketCounts,
+            sum(Sum) AS Sum,
+            sum(Count) AS Count,
+            min(MinDuration) AS Min,
+            max(MaxDuration) AS Max
+          FROM operation_latency_metrics_5_30_mv
+          WHERE Timestamp >= startDate AND Timestamp <= endDate
+            AND OrganizationID = '${params.organizationId}'
+            AND FederatedGraphID = '${params.graphId}'
+          GROUP BY Timestamp, OperationName, ExplicitBounds
+          ORDER BY Timestamp
+        )
+    `);
+    };
 
-    const prevP95 = this.client.query({
-      query: `histogram_quantile(0.95, sum(rate(cosmo_router_http_request_duration_milliseconds_bucket{${this.getQueryLabels(
-        params,
-      )}}[${range}h]] offset ${range}h)) by (le))`,
-    });
+    const p95 = queryLatency('0.95', start, end);
+    const prevP95 = queryLatency('0.95', prevStart, prevEnd);
 
     // get top 5 operations in last [range] hours
-    const top5 = this.client.query({
-      query: `topk(5, sum by (wg_operation_name) (rate(cosmo_router_http_request_duration_milliseconds_bucket{${this.getQueryLabels(
-        params,
-      )}}[${range}h])))`,
-    });
+    const queryTop5 = (quantile: string, start: number, end: number) => {
+      return this.chClient.queryPromise<{ name: string; value: string }>(`
+      WITH
+        toDateTime('${start}') AS startDate,
+        toDateTime('${end}') AS endDate
+      SELECT name, round(quantileDeterministic(0.5)(value, 1), 4) AS value FROM (
```

Comment:
Same here. We can rank the operation by P95.

Reply:
Same as above, get the median value of the aggregated time range